File: productExceptSelf.js
// Given an integer array (nums), return an array (answer) such that answer[i] is equal to the product of all the elements of nums except nums[i].

// The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

// You must write an algorithm that runs in O(n) time and without using the division operation.

// Example 1:
// Input: nums = [1,2,3,4]
// Output: [24,12,8,6]

// Example 2:
// Input: nums = [-1,1,0,-3,3]
// Output: [0,0,9,0,0]

function productExceptSelf(nums) {

    if(nums.length ===1){
        return [NaN]
    }
    let answer = []

    let leftProduct = 1
    for (let i = 0; i < nums.length; i++) {
        answer[i] = leftProduct
        console.log(answer)
        leftProduct *= nums[i]
    }

    let rightProduct = 1
    for (let i = nums.length - 1; i >= 0; i--) {
        answer[i] *= rightProduct
        rightProduct *= nums[i]
    }

    for(let i=0; i<answer.length; i++){
        if(answer[i]===-0){
            answer[i]=0
        }
    }
    return answer
}

module.exports = productExceptSelf;

File: productExceptSelf.py
# Given an integer array (nums), return an array (answer) such that answer[i] is equal to the product of all the elements of nums except nums[i].

# The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

# You must write an algorithm that runs in O(n) time and without using the division operation.

# Input: nums = [1,2,3,4]
# Example 1:
# Output: [24,12,8,6]

# Example 2:
# Input: nums = [-1,1,0,-3,3]
# Output: [0,0,9,0,0]

def product_except_self(nums):
    answer=[1]*len(nums)

    left_product=1
    for i in range(len(nums)):
        answer[i]=left_product
        left_product*=nums[i]

    right_product=1
    for i in range(len(nums)-1,-1,-1):
        answer[i]*=right_product
        right_product*=nums[i]
    
    print(answer) 

product_except_self([1,2,3,4])